creator-node/src/services/stateMachineManager/stateReconciliation/recoverOrphanedData.jobProcessor.ts

```diff
@@ -1,57 +1,303 @@
+      [QUEUE_NAMES.RECOVER_ORPHANED_DATA]: [
+        {
+          discoveryNodeEndpoint
+        }
+      ]
+    },
+    metricsToRecord: [
+      makeGaugeSetToRecord(
+        METRIC_NAMES.RECOVER_ORPHANED_DATA_WALLET_COUNTS_GAUGE,
+        walletsWithOrphanedData.length
+      ),
+      makeGaugeSetToRecord(
+        METRIC_NAMES.RECOVER_ORPHANED_DATA_SYNC_COUNTS_GAUGE,
+        requestsIssued
+      )
+    ]
+  }
+}
+
+/**
+ * Queries this node's db to find all users who have data on it and adds them to a redis set.
+ */
+const _storeWalletsOnThisNode = async () => {
+  await redisClient.del(WALLETS_ON_NODE_KEY)
+
+  // Make paginated SQL queries to find all wallets with data on this node
+  type WalletSqlRow = { walletPublicKey: string }
+  let prevWalletPublicKey = '0x'
+  const walletsOnThisNodeArr: string[] = []
+  let walletSqlRows: WalletSqlRow[] = []
+  do {
+    walletSqlRows = await models.CNodeUser.findAll({
+      attributes: ['walletPublicKey'],
+      order: [['walletPublicKey', 'ASC']],
+      where: {
+        walletPublicKey: {
+          [models.Sequelize.Op.gte]: prevWalletPublicKey
+        }
+      },
+      limit: NUM_USERS_PER_QUERY
+    })
+    walletSqlRows?.forEach((row) =>
+      walletsOnThisNodeArr.push(row.walletPublicKey)
+    )
+
+    // Move pagination cursor to the end
+    prevWalletPublicKey = walletsOnThisNodeArr.length
+      ? walletsOnThisNodeArr[walletsOnThisNodeArr.length - 1]
+      : '0x'
+  } while (
+    walletSqlRows?.length === NUM_USERS_PER_QUERY &&
+    prevWalletPublicKey !== '0x'
+  )
+
+  // Save the wallets as a redis set and return the set cardinality
+  const numWalletsOnNode = walletsOnThisNodeArr.length
+    ? await redisClient.sadd(WALLETS_ON_NODE_KEY, walletsOnThisNodeArr)
+    : 0
+  return numWalletsOnNode
+}
+
+/**
+ * Queries the given discovery node to find all users who have this content node as their primary or secondary.
+ * Adds them to a redis set.
+ */
+const _storeWalletsWithThisNodeInReplica = async (
+  discoveryNodeEndpoint: string
+) => {
+  await redisClient.del(WALLETS_WITH_NODE_IN_REPLICA_SET_KEY)
+
+  // Make paginated Discovery queries to find all wallets with this current node in their replica set (primary or secondary)
+  let prevUserId = 0
+  let batchOfUsers: StateMonitoringUser[] = []
+  const walletsWithNodeInReplicaSetArr: string[] = []
+  do {
+    // Get batch of users and all it to the array of all wallets
+    batchOfUsers = await getNodeUsers(
+      discoveryNodeEndpoint,
+      thisContentNodeEndpoint,
+      prevUserId,
+      NUM_USERS_PER_QUERY
+    )
+    batchOfUsers?.forEach((user) =>
+      walletsWithNodeInReplicaSetArr.push(user.wallet)
+    )
+
+    // Move pagination cursor to the end of the batch
+    prevUserId = batchOfUsers?.length
+      ? batchOfUsers[batchOfUsers.length - 1].user_id
+      : 0
+  } while (batchOfUsers?.length === NUM_USERS_PER_QUERY && prevUserId !== 0)
```

Comment:
hmm I'm not seeing how that would play well with pagination given that we're not getting every userId in the range. I think a do-while is very fitting here because the goal is to "do pagination while the cursor hasn't reached the end"

I also tested this manually for different edge cases to make sure it works. ex: batch size = 2 and there are 2 or 4 users, batch size = 1 and there are 2 users

Reply:
yeah i'm sure it works in the happy path
why wouldn't a for loop work with pagination? we first compute the max value and then cursor paginate through?

Reply:
how would it know when to stop paginating except in a while loop that checks when the cursor reaches the end (i.e., 0)? the userId wouldn't necessarily ever reach the max userId if it were just a for loop that was something like [i, maxUserId]

Reply:
not sure i follow this logic, can quickly chat on call

Reply:
just to write this out before our call
```
maxCNodeUserUUID = select max(cnodeUserUUID) from CNodeUsers
countCNodeUsers = select count(*) from cNodeUsers
for (let i = 0; i < countCNodeUsers; i += NUM_USERS_PER_QUERY) {
  ...cursor-based-paginated queries & logic
}
```

wdyt

Reply:
synced offline and added this for the local db query -- not currently possible for the Discovery query. also moved the saving-to-redis part into the pagination so it doesn't scale to eventually require keeping an array of millions of strings in memory before saving to redis

Reply:
good catch